sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,506 @@
+    else:
+        return 0
+
+
+def construct_c(num, den, x, poles, muls):
+    c = []
+    for pole, mul in zip(poles, muls):
+        c.append([])
+
+        # Case 3
+        if mul == 1:
+            # If multiplicity is 1, the coefficient to be added
+            # in the c-vector is 1 (no choice)
+            c[-1].extend([[1], [1]])
+
+        # Case 1
+        elif mul == 2:
+            # Find the coefficient of 1/(x - pole)**2 in the
+            # Laurent series expansion of a(x) about pole.
+            sgn, num1, den1 = (num*Poly((x - pole)**2, x, extension=True)).cancel(den)
+            r = (sgn*num1.subs(x, pole))/(den1.subs(x, pole))
+
+            # If multiplicity is 2, the coefficient to be added
+            # in the c-vector is c = (1 +- sqrt(1 + 4*r))/2
+            c[-1].extend([[(1 + sqrt(1 + 4*r))/2], [(1 - sqrt(1 + 4*r))/2]])
+
+        # Case 2
+        else:
+            # Generate the coefficients using the recurrence
+            # relation mentioned in (5.14) in the thesis (Pg 80)
+
+            # r_i = mul/2
+            ri = mul//2
+
+            # Find the Laurent series coefficients about the pole
+            ser = rational_laurent_series(num, den, x, pole, mul, 6)
+
+            # Start with an empty memo to store the coefficients
+            # This is for the plus case
+            temp = [0 for i in range(ri)]
+
+            # Base Case
+            temp[ri-1] = sqrt(ser[0])
+
+            # Iterate backwards to find all coefficients
+            for s in range(ri-1, 0, -1):
+                sm = 0
+                for j in range(s+1, ri):
+                    sm += temp[j-1]*temp[ri+s-j-1]
+                if s!= 1:
+                    temp[s-1] = (ser[mul-ri-s] - sm)/(2*temp[ri-1])
+
+            # Memo for the minus case
+            temp1 = [-x for x in temp]
+
+            # Find the 0th coefficient in the recurrence
+            temp[0] = (ser[mul-ri-s] - sm - ri*temp[ri-1])/(2*temp[ri-1])
+            temp1[0] = (ser[mul-ri-s] - sm  - ri*temp[ri-1])/(2*temp1[ri-1])
+
+            # Add both the plus and minus cases' coefficients
+            c[-1].extend([temp, temp1])
+    return c
+
+
+def construct_d(num, den, x, val_inf, mul):
+    N = -val_inf//2
+    ser = rational_laurent_series(num, den, x, oo, mul, 1)
+
+    # Case 4
+    if val_inf < 0:
+        temp = [0 for i in range(N+2)]
+        temp[N] = sqrt(ser[-mul - 1 + 2*N])
+        for s in range(N-1, -2, -1):
+            sm = 0
+            for j in range(s+1, N):
+                sm += temp[j]*temp[N+s-j]
+            if s != -1:
+                temp[s] = (ser[-mul - 1 + N+s] - sm)/(2*temp[N])
+        temp1 = [-x for x in temp]
+        # The third equation in Eq 5.15 of the thesis is WRONG!
+        # d_N must be replaced with N*d_N in that equation.
+        temp[-1] = (ser[-mul - 1 + N+s] - N*temp[N] - sm)/(2*temp[N])
+        temp1[-1] = (ser[-mul - 1 + N+s] - N*temp1[N] - sm)/(2*temp1[N])
+        d = [temp, temp1]
```

Comment:
When I looked at this to try and compare it with the thesis I found it quite confusing because of the variable names and the way that the indexing is done.

Firstly `temp` and `temp1` are not good variable names. I would call these `d1` and `d2` or `dplus` and `dminus` or something.

Secondly the indexing is confusing because of the `-mul-1` etc. I understand why you've done that because you have lists that have additional items in them but it would be better to use dicts instead of lists so that you can use whatever indices you want. Then the indices and variables should be made to match the equations in the thesis as closely as possible e.g. like:
```python
S[0] = sum(d[j]*d[N-1-j] for j in range(0, N-1+1))
d[-1] = (1 / (2*d[N])) * (a[N-1] - N*d[N] - S[0])
```